### PR TITLE
Fix accidental loss-of-precision for to_datetime(str, unit=...)

### LINKED ISSUE #57051: BUG: Inconsistent `to_datetime` behaviour
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> import pandas as pd

>>> pd.to_datetime("1704660000", unit="s", origin="unix")
<stdin>:1: FutureWarning: The behavior of 'to_datetime' with 'unit' when parsing strings is deprecated. In a future version, strings will be parsed as datetime strings, matching the behavior without a 'unit'. To retain the old behavior, explicitly cast ints or floats to numeric type before calling to_datetime.
Timestamp('2024-01-07 20:39:28')

>>> pd.to_datetime(1704660000, unit="s", origin="unix")
Timestamp('2024-01-07 20:40:00')
```

### Issue Description

Incorrect datetime object generated when unix timestamp is passed as string.

### Expected Behavior

The timestamp string should be converted to the correct datetime regarless of the `dtype` of the input (`string` or `int`). Or, raise error if `string` inputs are illegal.

The current depracation warning seems to imply that the behaviour is working as expected now, but won't be available in future versions.

### Installed Versions

```
[~] $ pip list | rg "pandas|numpy"
numpy                             1.26.1
pandas                            2.2.0
pandas-stubs                      2.1.4.231218
```

### PR DESCRIPTION
**Pull request by QuLogic**
In Pandas 1.5.3, the `float(val)` cast was inline to the `cast_from_unit` call in `array_with_unit_to_datetime`. This caused the intermediate (unnamed) value to be a Python float.

Since #50301, a temporary variable was added to avoid multiple casts, but with explicit type `cdef float`, which defines a _Cython_ float. This type is 32-bit, and causes a loss of precision, and a regression in parsing from 1.5.3.

Since `cast_from_unit` takes an `object`, not a more specific Cython type, remove the explicit type from the temporary `fval` variable entirely. This will cause it to be a (64-bit) Python float, and thus not lose precision.

Fixes #57051

- [x] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [n/a] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by WillAyd** [comment]
Can this be declared as double instead? Using a PyObject is unnecessary overhead

**Comment by QuLogic** [comment]
As noted, `cast_from_unit` takes an `object`, so that doesn't seem to save anything. Here's the diff of the generated when you add the type (vs having it commented out since that reduces line number changes in the diff):
```diff
--- untyped/pandas/_libs/tslib.pyx.c	2024-02-21 18:27:51.686387390 -0500
+++ typed/pandas/_libs/tslib.pyx.c	2024-02-21 18:24:14.824227898 -0500
@@ -25828,9 +25828,9 @@
   int __pyx_v_is_raise;
   PyArrayObject *__pyx_v_iresult = 0;
   PyDateTime_TZInfo *__pyx_v_tz = 0;
+  double __pyx_v_fval;
   PyObject *__pyx_v_result = NULL;
   PyObject *__pyx_v_val = NULL;
-  double __pyx_v_fval;
   PyObject *__pyx_v_err = NULL;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_iresult;
   __Pyx_Buffer __pyx_pybuffer_iresult;
@@ -25919,14 +25919,14 @@
  *         bint is_raise = errors == "raise"
  *         ndarray[int64_t] iresult
  *         tzinfo tz = None             # <<<<<<<<<<<<<<
- *         # double fval
+ *         double fval
  * 
  */
   __Pyx_INCREF(Py_None);
   __pyx_v_tz = ((PyDateTime_TZInfo *)Py_None);
 
   /* "pandas/_libs/tslib.pyx":280
- *         # double fval
+ *         double fval
  * 
  *     assert is_coerce or is_raise             # <<<<<<<<<<<<<<
  * 
@@ -32771,7 +32771,7 @@
  *     ndarray[object] values,
  *     str unit,
  */
-  __pyx_tuple__37 = PyTuple_Pack(13, __pyx_n_s_values, __pyx_n_s_unit, __pyx_n_s_errors, __pyx_n_s_i, __pyx_n_s_n, __pyx_n_s_is_coerce, __pyx_n_s_is_raise, __pyx_n_s_iresult, __pyx_n_s_tz, __pyx_n_s_result, __pyx_n_s_val, __pyx_n_s_fval, __pyx_n_s_err); if (unlikely(!__pyx_tuple__37)) __PYX_ERR(0, 239, __pyx_L1_error)
+  __pyx_tuple__37 = PyTuple_Pack(13, __pyx_n_s_values, __pyx_n_s_unit, __pyx_n_s_errors, __pyx_n_s_i, __pyx_n_s_n, __pyx_n_s_is_coerce, __pyx_n_s_is_raise, __pyx_n_s_iresult, __pyx_n_s_tz, __pyx_n_s_fval, __pyx_n_s_result, __pyx_n_s_val, __pyx_n_s_err); if (unlikely(!__pyx_tuple__37)) __PYX_ERR(0, 239, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__37);
   __Pyx_GIVEREF(__pyx_tuple__37);
   __pyx_codeobj__38 = (PyObject*)__Pyx_PyCode_New(3, 0, 0, 13, 0, CO_OPTIMIZED|CO_NEWLOCALS, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__37, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_home_elliott_code_pandas_pandas, __pyx_n_s_array_with_unit_to_datetime, 239, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__38)) __PYX_ERR(0, 239, __pyx_L1_error)
```

**Comment by WillAyd** [comment]
Fair point on the existing structure of the code. But if it works should still add it - generally the more typing in Cython the better. And if this gets refactored in the future the next developer won't have to think about it

**Review by WillAyd — APPROVED** [review]
lgtm

**Comment by WillAyd** [comment]
If you see us using float anywhere else probably worth changing to double everywhere in Cython (in a follow up PR)

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -1735,6 +1735,14 @@ def test_unit(self, cache):
         with pytest.raises(ValueError, match=msg):
             to_datetime([1], unit="D", format="%Y%m%d", cache=cache)
 
+    def test_unit_str(self, cache):
+        # GH 57051
+        # Test that strs aren't dropping precision to 32-bit accidentally.
+        with pytest.warns(FutureWarning):
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py:1741]:
```suggestion
        with tm.assert_produces_warning(FutureWarning):
```

**Comment by abhijeetbodas2001** [comment]
Thanks a lot for your investigation and fix for this @QuLogic!

File: pandas/tests/tools/test_to_datetime.py (hunk 1 above)
Comment (QuLogic) [reply pandas/tests/tools/test_to_datetime.py]: Done, and rebased.

File: pandas/tests/tools/test_to_datetime.py [hunk 2]
```diff
@@ -1735,6 +1735,14 @@ def test_unit(self, cache):
         with pytest.raises(ValueError, match=msg):
             to_datetime([1], unit="D", format="%Y%m%d", cache=cache)
 
+    def test_unit_str(self, cache):
+        # GH 57051
+        # Test that strs aren't dropping precision to 32-bit accidentally.
+        with tm.assert_produces_warning(FutureWarning):
+            res = pd.to_datetime(["1704660000"], unit="s", origin="unix")
+        expected = pd.to_datetime([1704660000], unit="s", origin="unix")
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py:1743]:
```suggestion
            res = to_datetime(["1704660000"], unit="s", origin="unix")
        expected = to_datetime([1704660000], unit="s", origin="unix")
```

File: doc/source/whatsnew/v2.2.1.rst [hunk 3]
```diff
@@ -50,6 +50,7 @@ Fixed regressions
 - Fixed regression in :meth:`Series.pct_change` raising a ``ValueError`` for an empty :class:`Series` (:issue:`57056`)
 - Fixed regression in :meth:`Series.to_numpy` when dtype is given as float and the data contains NaNs (:issue:`57121`)
 - Fixed regression in addition or subtraction of :class:`DateOffset` objects with millisecond components to ``datetime64`` :class:`Index`, :class:`Series`, or :class:`DataFrame` (:issue:`57529`)
+- Fixed regression in precision of :func:`to_datetime` with string and ``unit`` input (:issue:`57051`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.1.rst:53]: Could you move this to `v2.2.2.rst`?

Comment (QuLogic) [reply doc/source/whatsnew/v2.2.1.rst]: Done.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @QuLogic

**Pull request merged**